### PR TITLE
Whitelist Should Not Accept New Fee Currencies Without An Oracle Price

### LINKED ISSUE #7315: Whitelist should not accept new fee currencies without an oracle price
### Challenge Description
FeeCurrencies are a unique feature of the Celo platform that allow the payment of gas in non-native tokens, in Celo's case stable value assets which will yield a better UX for users. This bounty makes the addition of those feeCurrencies safer by requiring that `SortedOracles#medianRate` with the currency actually returns a valid value (which is what `celo-blockchain` uses to sort transaction by gasPrice)

**The bounty payout amount will be for $200 worth of CELO at the time of completion.**

### Submission Requirements
- Submit a valid PR that passes all CI checks
- Adjust existing tests
- Add unit tests that check the invariant that feeCurrencies should not succeed if an oracle price is not existent on `SortedOracle` + succeed when oracle prices are reported
- Address PR reviews by the core developers
- Have the PR be merged

### Prerequisites
- Basic Solidity Knowledge
- Minimal Familiarity with Celo Core Contracts

### Time Expectation
- 4 hours

### Resources
- https://docs.celo.org/celo-codebase/protocol/stability
- https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/Exchange.sol#L20:L20
- https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/SortedOracles.sol#L232:L232
- https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/common/FeeCurrencyWhitelist.sol#L26:L26
- [Oracle mocking medianRate example](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/test/stability/exchange.ts#L134:L134)

### Expected Behavior

FeeCurrencyWhitelist.addToken reverts if no oracle rates are found for that token on `SortedOracles`. This is needed for celo-blockchain to sort transactions by gas price across different currencies. 

### Current Behavior

No such check, and so the governance approvers and voters need to exercise caution when adding fee currencies to the whitelist.

### PR DESCRIPTION
**Pull request by karbafoo**
### Description

Updated addToken function to reject tokens without a valid oracle price

### Tested

unit tests added

### Related issues

- Fixes #7315

### Backwards compatibility

initialize method needs a register address. addToken change doesnt effect other features

### CODE REVIEW

File: packages/protocol/test/common/gascurrencywhitelist.ts [hunk 1]
```diff
@@ -1,40 +1,88 @@
...
+            await mockSortedOracles.setMedianRate(stableToken.address, stableAmountForRate)
+            await mockSortedOracles.setMedianTimestampToNow(stableToken.address)
+            await mockSortedOracles.setNumRates(stableToken.address, 2)
+            await feeCurrencyWhitelist.addToken(stableToken.address)
+            const tokens = await feeCurrencyWhitelist.getWhitelist()
+            assert.sameMembers(tokens, [stableToken.address])
+        })
+
+        it('should not allow a non-owner to add a token', async () => {
+            await assertRevert(feeCurrencyWhitelist.addToken(stableToken.address, { from: nonOwner }))
```
Comment (nambrot) [inline packages/protocol/test/common/gascurrencywhitelist.ts]: Think about making this test specific to the provision of the non-owner. Right now this test would fail even with the owner as no token price was reported. I'd recommend making a new describe/context block that sets the token price in a `beforeEach`

File: packages/protocol/test/common/gascurrencywhitelist.ts [hunk 2]
```diff
@@ -1,40 +1,88 @@
...
-  const aTokenAddress = '0x000000000000000000000000000000000000ce10'
+    const nonOwner = accounts[1]
 
-  const nonOwner = accounts[1]
+    const decimals = 18
+    const SECONDS_IN_A_WEEK = 604800
+    const goldAmountForRate = new BigNumber('0x10000000000000000')
+    const stableAmountForRate = new BigNumber(2).times(goldAmountForRate)
+    beforeEach(async () => {
+        stableToken = await StableToken.new()
```
Comment (nambrot) [inline packages/protocol/test/common/gascurrencywhitelist.ts]: This would be optional, but could you consider using `MockStableToken` instead of `StableToken`, then you would having to provide these initialization parameters

File: packages/protocol/test/common/gascurrencywhitelist.ts [hunk 3]
```diff
@@ -1,40 +1,88 @@
...
-      assert.sameMembers(tokens, [aTokenAddress])
+        it('should not be callable again', async () => {
+            await assertRevert(feeCurrencyWhitelist.initialize(registry.address))
+        })
     })
 
-    it('should not allow a non-owner to add a token', async () => {
-      await assertRevert(feeCurrencyWhitelist.addToken(aTokenAddress, { from: nonOwner }))
+    describe('#addToken()', () => {
+        it('should not allow to add a token with an invalid oracle price', async () => {
```
Comment (nambrot) [inline packages/protocol/test/common/gascurrencywhitelist.ts]: I might rephrase this to 
```suggestion
        it('should not allow to add a token without an oracle price', async () => {
```

**Review by nambrot — COMMENTED** [review]
Just a few minor comments!

**Comment by karbafoo** [comment]
should I make the list unique ?

**Comment by nambrot** [comment]
If you mean the fee currency whitelist, I think that is a good point, but I wouldn't worry about it in this PR

File: packages/protocol/test/common/gascurrencywhitelist.ts [hunk 4]
```diff
@@ -1,88 +1,75 @@
...
+    it('should not be callable again', async () => {
+      await assertRevert(feeCurrencyWhitelist.initialize(registry.address))
     })
+  })
 
-    describe('#addToken()', () => {
-        it('should not allow to add a token with an invalid oracle price', async () => {
-            await assertRevert(feeCurrencyWhitelist.addToken(stableToken.address))
-        })
+  describe('#addToken() with an invalid oracle price', () => {
```
Comment (nambrot) [inline packages/protocol/test/common/gascurrencywhitelist.ts]: Sorry this might seem like a nit, but our general convention is to have one top level `describe` for the function in question (in this case `addToken`) and then have nested describe's that express the different cases

**Review by nambrot — CHANGES_REQUESTED** [review]
Just one more point and then we should be good!

File: packages/protocol/test/common/gascurrencywhitelist.ts (hunk 4 above)
Comment (karbafoo) [reply packages/protocol/test/common/gascurrencywhitelist.ts]: my bad sorry

File: packages/protocol/test/common/gascurrencywhitelist.ts [hunk 5]
```diff
@@ -1,18 +1,41 @@
...
   const nonOwner = accounts[1]
 
+  const goldAmountForRate = new BigNumber('0x10000000000000000')
+  const stableAmountForRate = new BigNumber(2).times(goldAmountForRate)
   beforeEach(async () => {
-    feeCurrencyWhitelist = await FeeCurrencyWhitelist.new()
-    await feeCurrencyWhitelist.initialize()
+    mockStableToken = await MockStableToken.new()
+    registry = await Registry.new()
+    // await registry.setAddressFor(CeloContractName.StableToken, mockStableToken.address)
```
Comment (nambrot) [inline packages/protocol/test/common/gascurrencywhitelist.ts]: Remove?
```suggestion
```

File: packages/protocol/test/common/gascurrencywhitelist.ts [hunk 6]
```diff
@@ -22,19 +45,34 @@ contract('FeeCurrencyWhitelist', (accounts: string[]) => {
...
     })
   })
-
   describe('#addToken()', () => {
-    it('should allow the owner to add a token', async () => {
-      await feeCurrencyWhitelist.addToken(aTokenAddress)
-      const tokens = await feeCurrencyWhitelist.getWhitelist()
-      assert.sameMembers(tokens, [aTokenAddress])
+    describe('when token has an invalid oracle price', () => {
+      it('should not allow to add a token with an invalid oracle price', async () => {
```
Comment (nambrot) [inline packages/protocol/test/common/gascurrencywhitelist.ts]:
```suggestion
      it('should not allow to add a token without a valid oracle price', async () => {
```

File: packages/protocol/contracts/common/FeeCurrencyWhitelist.sol [hunk 7]
```diff
@@ -4,26 +4,40 @@ import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
...
    * @dev Add a token to the whitelist
    * @param tokenAddress The address of the token to add.
    */
   function addToken(address tokenAddress) external onlyOwner {
+    uint256 rateNumerator;
+    uint256 rateDenominator;
+    (rateNumerator, rateDenominator) = ISortedOracles(
+      registry.getAddressForOrDie(SORTED_ORACLES_REGISTRY_ID)
+    )
+      .medianRate(tokenAddress);
```
Comment (yorhodes) [inline packages/protocol/contracts/common/FeeCurrencyWhitelist.sol]:
```suggestion
    uint256 rateNumerator;
    uint256 rateDenominator;
    (rateNumerator, rateDenominator) = getSortedOracles().medianRate(tokenAddress);
```

File: packages/protocol/contracts/common/FeeCurrencyWhitelist.sol [hunk 8]
```diff
@@ -4,26 +4,40 @@ import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
 
 import "./interfaces/IFeeCurrencyWhitelist.sol";
 
-import "../common/Initializable.sol";
+import "../common/InitializableV2.sol";
```
Comment (yorhodes) [inline packages/protocol/contracts/common/FeeCurrencyWhitelist.sol]:
```suggestion
import "./InitializableV2.sol";
```

File: packages/protocol/contracts/common/FeeCurrencyWhitelist.sol [hunk 9]
```diff
@@ -4,26 +4,40 @@ import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
...
 
+import "../common/UsingRegistry.sol";
```
Comment (yorhodes) [inline packages/protocol/contracts/common/FeeCurrencyWhitelist.sol]:
```suggestion
import "./UsingRegistry.sol";
```

**Review by yorhodes — COMMENTED** [review]
waiting for a few versioning changes to approve cc @nambrot

File: packages/protocol/contracts/common/FeeCurrencyWhitelist.sol [hunk 10]
```diff
@@ -4,26 +4,37 @@ import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
...
+  constructor(bool test) public InitializableV2(test) {}
+
   /**
    * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param registryAddress Address of the Registry contract.
    */
-  function initialize() external initializer {
+  function initialize(address registryAddress) external initializer {
     _transferOwnership(msg.sender);
+    setRegistry(registryAddress);
```
Comment (yorhodes) [inline packages/protocol/contracts/common/FeeCurrencyWhitelist.sol:25]: By introducing new storage, we imply that `FeeCurrencyWhitelist` will have a new storage contract deployed and initialized.

See https://docs.celo.org/community/release-process/smart-contracts#versioning

If we were to redeploy `FeeCurrencyWhitelist`, we would also need to migrate the existing whitelist to the new address atomically in the `Governance` proposal. It was an oversight not to deploy this contract with `UsingRegistry` in the first place, but I think the upgrade complexity is not worth the syntactical benefits we gain. 

My recommendation would be to give `FeeCurrencyWhitelist` special knowledge of the (likely) indefinitely static registry address `0x0..ce10` in its bytecode. 

cc @nambrot

**Review by yorhodes — CHANGES_REQUESTED** [review]
I missed that this was introducing a new storage variable which would necessitate initialization.

**Comment by yorhodes** [comment]
> I missed that this was introducing a new storage variable which would necessitate initialization.

I am working on a resolution for this problem. @karbafoo there is no further action needed from you at this time and I will make sure that your bounty assignment doesn't expire.
